src/core/file-impl.hh

```diff
@@ -154,32 +154,50 @@ private:
      *         pos is at or beyond EOF.
      * @throw appropriate exception in case of I/O error.
      */
-    future<temporary_buffer<uint8_t>>
-    read_maybe_eof(uint64_t pos, size_t len, const io_priority_class& pc, io_intent* intent);
+    future<temporary_buffer<uint8_t>> read_maybe_eof(uint64_t pos, size_t len, internal::maybe_priority_class_ref pc, io_intent* intent);
+    future<size_t> read_dma_one(uint64_t pos, void* buffer, size_t len, internal::maybe_priority_class_ref pc, io_intent* intent) noexcept {
+        return read_dma(pos, buffer, len, pc.pc, intent);
+    }
 
 protected:
-    future<size_t> do_write_dma(uint64_t pos, const void* buffer, size_t len, const io_priority_class& pc, io_intent* intent) noexcept;
-    future<size_t> do_write_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc, io_intent* intent) noexcept;
-    future<size_t> do_read_dma(uint64_t pos, void* buffer, size_t len, const io_priority_class& pc, io_intent* intent) noexcept;
-    future<size_t> do_read_dma(uint64_t pos, std::vector<iovec> iov, const io_priority_class& pc, io_intent* intent) noexcept;
-    future<temporary_buffer<uint8_t>> do_dma_read_bulk(uint64_t offset, size_t range_size, const io_priority_class& pc, io_intent* intent) noexcept;
+    future<size_t> do_write_dma(uint64_t pos, const void* buffer, size_t len, internal::maybe_priority_class_ref pc, io_intent* intent) noexcept;
+    future<size_t> do_write_dma(uint64_t pos, std::vector<iovec> iov, internal::maybe_priority_class_ref pc, io_intent* intent) noexcept;
+    future<size_t> do_read_dma(uint64_t pos, void* buffer, size_t len, internal::maybe_priority_class_ref pc, io_intent* intent) noexcept;
+    future<size_t> do_read_dma(uint64_t pos, std::vector<iovec> iov, internal::maybe_priority_class_ref pc, io_intent* intent) noexcept;
+    future<temporary_buffer<uint8_t>> do_dma_read_bulk(uint64_t offset, size_t range_size, internal::maybe_priority_class_ref pc, io_intent* intent) noexcept;
```

Comment:
Since these are exposed to external file_impl implementations, please check that scylladb still compiles. I expect it should.

Reply:
Scylla still compiles

Reply:
@xemul hi Pavel, sorry for commenting in a closed pull request. do you have any plan on s/io_priority/scheduling_group/ in scylladb? usually, i compile scylladb without `-Wno-error=deprecated-declarations`. and i hope we can drop it in `configure.py` as well. also from the practical perspective, we need to address the problem explained in https://github.com/scylladb/seastar/issues/1069#issue-1241298126 in scylla sooner or latter.

i take a quick look at the code paths where io_priority is used, and found the task non-trivial. in addition to replacing all occurrences of io priorities with scheduling groups, we also need to create and destroy them. but unlike `::io_priority_class::register_one`, `create_scheduling_group()` is a coroutine, so we cannot create them in a plain constructor like `priority_manager::priority_manager`, not to mention having a `static thread_local priority_manager`. probably we can create a sharded service in `main()` which provides the scheduling_groups for different i/o ops, or we should instead leave it to the tasks which actually issue io and let them to use dedicated a scheduling_group in certain cases where the current one does not suit the needs.

Reply:
ahh, we do have different scheduling_groups in `database_config`, so what we need is probably to audit all the code paths and populate these sg down to where the io ops are issued if the current sg does not suit.

Reply:
@tchaikov , my plan (nearest, Avi had already refreshed seastar in Scylla) is to remove the io_priority_class at all from Scylla code and :crossed_fingers: hope that whatever sched group happens to be current in each IO would be "correct" one. Likely we'll step on some bugs, but I guess there's no more reliable way to spot them.

Also, I do know several IO places in Scylla code that occur with default io-class, while should go in some other one, and those places _do_ have correct current sched groups, so these places will be auto-fixed

Reply:
thank you @xemul, sounds great!  much appreciated! i spent a while the other day to do a mechanical compile-search-replace loop, but ended up giving it up. there were just too many of them, and the more places i changed the more scared -- i was afraid that i was introducing "some bugs" without being aware of them. as you put, 

>  there's no more reliable way to spot them.

anyway, thank you! sometimes, the harmless warning messages are kind of distracting =)

Reply:
There's just one mismatch in scylladb where scheduling groups and io_priority_class don't match, and that's commitlog. We'll need to create a commitlog scheduling group.